### PR TITLE
Implement gather primitive for ProcessGroupNCCL

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #71623
* #70029
* __->__ #66745

This PR implement NCCL gather and add gather to ProcessGroupNCCL using nccl send/recv api.

NCCL doesn’t directly provide primitives for gather, so we need to be implemented on top of NCCL’s send/recv API.
1. In ProcessGroupNCCL.cpp, the outputTensors are first flattened, then inputTensors and outputFlattened are passed by the collective class to gather() function in nccl.cpp.
1. In nccl.cpp, gather is implemented using ncclSend/ncclRecv: all the ranks send inputTensor to the root rank, and the root rank uses a for loop to receive these inputTensors.

Differential Revision: [D29616361](https://our.internmc.facebook.com/intern/diff/D29616361/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29616361/)!

### CODE REVIEW

File: test/distributed/test_c10d_nccl.py [hunk 1]
```diff
@@ -464,6 +464,215 @@ def allgather_base(output_t, input_t):
             # fails the check because the dtype is different
             allgather_base(output_t, tensor)
 
+    @requires_nccl()
+    def test_gather_ops(self):
```
[the remark this answers is not in the document]
Comment (wanchaol) [reply test/distributed/test_c10d_nccl.py:497]: I have tried merging these two actually, one problem is that ProcessGroupNCCL and ProcessGroupGloo behavior is different in terms of result for example (see https://github.com/pytorch/pytorch/issues/68478), I will work on making them consistent as a follow up diff, then merging the tests together

File: torch/csrc/cuda/nccl.cpp [hunk 2]
```diff
@@ -811,6 +811,47 @@ void recv(
...
+  auto type = to_nccl_data_type(inputs);
+  const auto* sendbuff = reinterpret_cast<char*>(inputs.data_ptr());
+
+  NCCL_CHECK(ncclGroupStart());
+  NCCL_CHECK(ncclSend(sendbuff, count, type, root, comm, stream));
+  if (cur_rank == root)
+  {
+    for (int r = 0; r < numranks; r++)
+    {
+      auto* recvbuff =  reinterpret_cast<char*>(outputs[r].data_ptr());
```
Comment (pritamdamania87) [inline torch/csrc/cuda/nccl.cpp:840]: nit: outputs[r] needs to create a new Tensor object (not necessarily a copy) and might have some minor overhead.. Was wondering if this can just be done using output.data_ptr().

File: torch/csrc/cuda/nccl.cpp [hunk 3]
```diff
@@ -811,6 +811,47 @@ void recv(
...
+  int numranks, cur_rank;
+  NCCL_CHECK(ncclCommCount(comm, &numranks));
+  NCCL_CHECK(ncclCommUserRank(comm, &cur_rank));
+
+  size_t count = inputs.numel();
+  auto type = to_nccl_data_type(inputs);
+  const auto* sendbuff = reinterpret_cast<char*>(inputs.data_ptr());
+
+  NCCL_CHECK(ncclGroupStart());
+  NCCL_CHECK(ncclSend(sendbuff, count, type, root, comm, stream));
```
Comment (pritamdamania87) [inline torch/csrc/cuda/nccl.cpp:835]: The root rank shouldn't have to send to self?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 4]
```diff
@@ -2142,10 +2142,95 @@ void ProcessGroupNCCL::groupEnd() {
...
+
+  // @lint-ignore CLANGTIDY
+  auto tensor = inputTensors.back();
+  RECORD_PARAM_COMMS(
+      rank_,                    // rank
+      "gather",             // colName
+      tensor.numel(),           // inSize
+      tensor.numel() *          // outSize
+        this->getSize(),        // dType
+      tensor.scalar_type(),
```
Comment (fduwjj) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:2166]: Nit: looks like comment and code are not aligned? `// outSize` should be in L2165, `// dType` should be in L 2166?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 5]
```diff
@@ -2142,10 +2142,95 @@ void ProcessGroupNCCL::groupEnd() {
...
+    const auto& options = inputTensors[0].options();
+    const auto& sizes = inputTensors[0].sizes();
+    assertTypeAndSizesMatch(invalidArgument, outputTensors[0], options, sizes);
+  } else {
+    // if not in the root rank, initialize outputFlattened as empty place holder
+    if (outputTensors.size() != 0) {
+      invalidArgument("requires empty output on non-root");
+    }
+
+    for(int i = 0; i < inputTensors.size(); ++ i) {
```
Comment (fduwjj) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:2196]: Nit: i++? (or ++i?)

File: torch/csrc/cuda/nccl.cpp (hunk 2 above)
Comment (pritamdamania87) [reply torch/csrc/cuda/nccl.cpp]: The `[]` operator has a bunch of overhead and creates a new Tensor object: https://codebrowser.bddppq.com/pytorch/pytorch/aten/src/ATen/TensorOperators.h.html#_ZNK2at6TensorixEN3c106ScalarE. It might be slightly faster to just use `.data_ptr()` directly.

**Review by pritamdamania87 — APPROVED** [review]

File: torch/csrc/cuda/nccl.cpp (hunk 2 above)
Comment (pritamdamania87) [reply torch/csrc/cuda/nccl.cpp]: Although, I feel this is probably fine for now since directly manipulating `data_ptr()` would involve getting strides and dtype etc and some pointer arithmetic.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 6]
```diff
@@ -2151,7 +2278,7 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::gather(
   check_gpu_tensors(inputTensors);
   assertSingleElementInput(invalidArgument, inputTensors);
 
-  std::vector<at::Tensor> outputFlattened;
+  // std::vector<at::Tensor> outputFlattened;
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:2281]: nit: remove completely?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 7]
```diff
@@ -1438,6 +1438,116 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
...
+  auto work = initWork(
+      devices,
+      rank_,
+      opType,
+      can_profile ? profilingTitle : nullptr,
+      can_profile ? c10::optional<std::vector<at::Tensor>>(tensor)
+                  : c10::nullopt);
+
+  // Store references to outputs to be used by WorkNCCL::result and operator<<.
+  if (opType == OpType::GATHER) {
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1473]: Why do we have this conditional here?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 8]
```diff
@@ -1438,6 +1438,116 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
   return work;
 }
 
+template <typename Fn, typename PreProcess, typename PostProcess>
+c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::oneAndAll(
+    std::vector<at::Tensor>& tensor,
+    std::vector<std::vector<at::Tensor>>& tensor_list,
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1444]: Do we really need a new method like this that has a lot of similarities to `collective`? What if we just make `collective` take a `vector<vector<Tensor>>` instead of just `vector<Tensor>`? Then modify current code that passes in a `vector<Tensor>` to pass in a `vector<vector<Tensor>>` of size 1?

Comment (wanchaol) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1444]: Yeah I tried to consolidate all into `collective` before introducing this oneAndAll, failed mainly because for `gather` and `scatter`, the inputs and outputs are different (they are swapped), so need someway to distinguish it (i.e. we only can record input stream on one rank for `scatter`, not all ranks as its only available on rootRank, same for `gather`), so it was pretty hard to generalize. Let me try again and see if this could work

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 7 above)
Comment (wanchaol) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1473]: This is because for Gather and Scatter, we are actually defining `outputs` differently, for `gather`, `outputs` is `tensor_list`, while for `scatter`, `outputs` is defined as `tensor`

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 8 above)
Comment (wanchaol) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp]: I actually found a way to work around this issue, by not introducing `oneAndAll`, and still reuse `collective` to do the job, the trick is to ignore the inputs/outputs specified in `fn` and use function capture to deal with `tensor` and `tensor_list` directly. Updated the PRs to reflect the changes. Let me know if this works :)

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 9]
```diff
@@ -2140,10 +2140,74 @@ void ProcessGroupNCCL::groupEnd() {
...
+
+    const auto& options = inputTensors[0].options();
+    const auto& sizes = inputTensors[0].sizes();
+    assertTypeAndSizesMatch(invalidArgument, outputTensors[0], options, sizes);
+  } else {
+    // if not in the root rank, initialize outputFlattened as empty place holder
+    if (outputTensors.size() != 0) {
+      invalidArgument("requires empty output on non-root");
+    }
+    outputTensors.emplace_back();
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:2188]: nit: probably not a good idea to modify the variable passed in. Maybe just create an empty vector<at:Tensor> and pass that in instead?

**Review by pritamdamania87 — APPROVED** [review]
